packages/contracts-bedrock/src/L1/SystemConfigInterop.sol

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+import { Constants } from "src/libraries/Constants.sol";
+import { OptimismPortalInterop as OptimismPortal } from "src/L1/OptimismPortalInterop.sol";
+import { GasPayingToken } from "src/libraries/GasPayingToken.sol";
+import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import { SystemConfig } from "src/L1/SystemConfig.sol";
+import { ConfigType } from "src/L2/L1BlockInterop.sol";
+
+/// @notice Error thrown when an address has no code.
+error NoCode(address);
+
+/// @title SystemConfigInterop
+/// @notice The SystemConfig contract is used to manage configuration of an Optimism network.
+///         All configuration is stored on L1 and picked up by L2 as part of the derviation of
+///         the L2 chain.
+contract SystemConfigInterop is SystemConfig {
+    /// @notice Modifier to check if an address has code.
+    /// @param _address Address to check.
+    /// @param _check Boolean to determine if the check should be performed.
+    modifier mustHaveCode(address _address, bool _check) {
+        if (_check) {
+            assembly {
+                let size := extcodesize(_address)
+                // Revert if the code size is 0.
+                if iszero(size) {
+                    mstore(0x00, 0x674520b6) // 0x674520b6 is the 4-byte selector of "NoCode(address)"
+                    revert(0x1C, 0x04) // returns the stored 4-byte selector from above
+                }
+            }
+        }
+        _;
+    }
+
+    /// @custom:semver 2.2.0+interop
+    function version() public pure override returns (string memory) {
+        return string.concat(super.version(), "+interop");
+    }
+
+    /// @notice Internal setter for the gas paying token address, includes validation.
+    ///         The token must not already be set and must be non zero and not the ether address
+    ///         to set the token address. This prevents the token address from being changed
+    ///         and makes it explicitly opt-in to use custom gas token. Additionally,
+    ///         OptimismPortal's address must be non zero, since otherwise the call to set the
+    ///         config for the gas paying token to OptimismPortal will fail.
+    /// @param _token Address of the gas paying token.
+    function _setGasPayingToken(address _token)
+        internal
+        override
+        mustHaveCode(optimismPortal(), _token != address(0) && _token != Constants.ETHER && !isCustomGasToken())
```

Comment:
I don't think it makes sense to make `mustHaveCode` into a modifier, it makes it complex to reason about with the second argument, the abstraction isn't ideal for testing as now you need 2 test cases. I recommend getting rid of the modifier and just inlining the codesize check in this function